src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TaskStatusCenter;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    [Export(typeof(DiagnosticProgressReporter))]
+    internal sealed class DiagnosticProgressReporter
+    {
+        private static readonly TimeSpan s_minimumInterval = TimeSpan.FromMilliseconds(200);
+
+        private readonly IVsTaskStatusCenterService _taskCenterService;
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly TaskHandlerOptions _options;
+
+        // these fields are never accessed concurrently
+        private TaskCompletionSource<VoidResult> _currentTask;
+        private object _lastReportedDocumentOrProject;
+        private DateTimeOffset _lastTimeReported;
+
+        // this is only field that is shared between 2 events streams (IDiagnosticService and ISolutionCrawlerProgressReporter)
+        // and can be called concurrently.
+        private volatile ITaskHandler _taskHandler;
+
+        [ImportingConstructor]
+        public DiagnosticProgressReporter(
+            SVsServiceProvider serviceProvider,
+            IDiagnosticService diagnosticService,
+            VisualStudioWorkspace workspace)
+        {
+            _lastTimeReported = DateTimeOffset.UtcNow;
+
+            _taskCenterService = (IVsTaskStatusCenterService)serviceProvider.GetService(typeof(SVsTaskStatusCenterService));
```

Comment:
We quite frequently see deadlocks attributable to folks doing STA things in ImportingConstructors. Are you *sure* this MEF part is only activated while on the UI thread? Unless you can be sure of now and future use of this MEF part, I recommend you defer all STA operations to sometime after MEF activation is done to avoid deadlocks.

Reply:
yes, it is called in UI thread.

this ensures it

   protected override void LoadComponentsInUIContext()
          {		          {
              // we need to load it as early as possible since we can have errors from		              // we need to load it as early as possible since we can have errors from
              // package from each language very early		              // package from each language very early
 +            this.ComponentModel.GetService<DiagnosticProgressReporter>();
              this.ComponentModel.GetService<VisualStudioDiagnosticListTable>();		              this.ComponentModel.GetService<VisualStudioDiagnosticListTable>();